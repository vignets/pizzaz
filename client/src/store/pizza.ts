import { API_URL } from "@/config/env-vars";
import { Pizza } from "@/types";
import { atom } from "jotai";

export const prepareData = async () => {
  const response = await fetch(`${API_URL}/pizzas.json`);
  const data = (await response.json()) as Pizza[];
  return data.map((user) => ({ ...user, quantity: 0 }));
};

export const pizzasAtom = atom<Pizza[]>([]);
