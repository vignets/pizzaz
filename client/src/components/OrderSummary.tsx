import { pizzasAtom } from "@/store";
import { Button } from "@mantine/core";
import { useAtom } from "jotai";
import React from "react";
import { ArrowRightIcon } from "@heroicons/react/outline";

export default function OrderSummary() {
  const [pizzas, setPizzas] = useAtom(pizzasAtom);
  const orderedItems = pizzas.filter((item) => item.quantity > 0);
  if (orderedItems.length === 0) return null;
  const total = pizzas.reduce(function (previousValue, currentValue) {
    return previousValue + currentValue.quantity * currentValue.price;
  }, 0);

  return (
    <div className="fixed left-1/2 -translate-x-1/2 bottom-0 w-full max-w-4xl ">
      <div className="bg-indigo-600 text-white h-full mx-4 sm:mx-6 lg:mx-8 rounded-t-2xl px-6 py-4">
        <div>
          <h2 className=" font-bold text-2xl">Order Summary</h2>
          <div className="flex justify-between items-end">
            <div className="flex items-baseline space-x-2">
              <p className="text-sm font-semibold">Total:</p>
              <p className="text-2xl font-extrabold ">
                <span className="font-medium">$</span>
                {total.toFixed()}
                <span className="text-lg">
                  {"."}
                  {total.toFixed(2).split(".")[1]}
                </span>
              </p>
            </div>
            <Button variant="default" className="text-gray-800">
              Place Order <ArrowRightIcon className="w-5 h-5 ml-1" />
            </Button>
          </div>
        </div>
      </div>
    </div>
  );
}
