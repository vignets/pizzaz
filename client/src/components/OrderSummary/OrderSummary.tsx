import { pizzasAtom } from "@/store";
import { Button } from "@mantine/core";
import { useAtomValue } from "jotai";
import React from "react";
import { ArrowRightIcon, CheckIcon } from "@heroicons/react/outline";
import { API_URL } from "@/config/env-vars";
import { showNotification } from "@mantine/notifications";

export default function OrderSummary() {
  const pizzas = useAtomValue(pizzasAtom);
  const orderedItems = pizzas.filter((item) => item.quantity > 0);
  if (orderedItems.length === 0) return null;
  const total = pizzas.reduce(function (previousValue, currentValue) {
    return previousValue + currentValue.quantity * currentValue.price;
  }, 0);

  return (
    <div
      className="fixed left-1/2 -translate-x-1/2 bottom-0 w-full max-w-4xl"
      data-testid="order-summary"
    >
      <div className="h-full bg-indigo-600 text-white mx-4 sm:mx-6 lg:mx-8 rounded-t-2xl px-6 py-4">
        <div>
          <h2 className="font-bold text-2xl">Order Summary</h2>
          <div className="flex justify-between items-end">
            <div className="flex items-baseline space-x-2">
              <p className="text-sm font-semibold">Total:</p>
              <p className="text-2xl font-extrabold " data-testid="total-price">
                <span className="font-medium">$</span>
                {total.toFixed()}
                <span className="text-lg">
                  {"."}
                  {total.toFixed(2).split(".")[1]}
                </span>
              </p>
            </div>
            <Button
              variant="default"
              className="text-gray-800"
              onClick={() => {
                fetch(`${API_URL}/order.json`)
                  .then((res) => res.json())
                  .then((data) => {
                    if (!data.success) throw Error("Server Error.");
                    showNotification({
                      title: "Order Placed",
                      message: `Your order will be delivered in ${
                        data.deliveryTime / 1000 / 60
                      } minutes.`,
                      autoClose: 5000,
                      icon: <CheckIcon className="w-6 h-6" />,
                    });
                  })
                  .catch((err) =>
                    showNotification({
                      title: "Something went wrong ðŸ˜•.",
                      message: err?.message,
                      autoClose: 5000,
                    }),
                  );
              }}
            >
              Place Order <ArrowRightIcon className="w-5 h-5 ml-1" />
            </Button>
          </div>
        </div>
      </div>
    </div>
  );
}
