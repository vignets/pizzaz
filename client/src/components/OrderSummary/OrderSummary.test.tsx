import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import OrderSummary from "./OrderSummary";
import { pizzasAtom } from "@/store";
import { Provider } from "jotai";
import { NotificationsProvider } from "@mantine/notifications";

describe("Order Summary", () => {
  it("does not render counter", async () => {
    render(<OrderSummary />);

    const heading = screen.queryByText(/order summary/i);
    expect(heading).not.toBeInTheDocument();
  });

  it("renders counter", async () => {
    const { container } = render(
      <NotificationsProvider position="top-right">
        <Provider
          initialValues={
            [
              [
                pizzasAtom,
                [
                  {
                    name: "Margherita",
                    price: 5,
                    ingredients: ["tomato"],
                    quantity: 2,
                  },
                ],
              ],
            ] as const
          }
        >
          <OrderSummary />
        </Provider>
      </NotificationsProvider>,
    );

    const heading = screen.queryByText(/order summary/i);
    expect(heading).toBeInTheDocument();

    expect(screen.getByTestId("total-price")).toHaveTextContent("$10.00");

    const order = screen.getByRole("button", { name: /place order/i });

    fireEvent.click(order);

    await waitFor(() => {
      screen.getByText(/order placed/i);
    });

    expect(container).toMatchSnapshot();
  });
});
