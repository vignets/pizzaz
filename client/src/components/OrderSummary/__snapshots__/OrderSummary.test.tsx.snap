// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Summary renders counter 1`] = `
<div>
  <div
    class="fixed left-1/2 -translate-x-1/2 bottom-0 w-full max-w-4xl"
    data-testid="order-summary"
  >
    <div
      class="h-full bg-indigo-600 text-white mx-4 sm:mx-6 lg:mx-8 rounded-t-2xl px-6 py-4"
    >
      <div>
        <h2
          class="font-bold text-2xl"
        >
          Order Summary
        </h2>
        <div
          class="flex justify-between items-end"
        >
          <div
            class="flex items-baseline space-x-2"
          >
            <p
              class="text-sm font-semibold"
            >
              Total:
            </p>
            <p
              class="text-2xl font-extrabold "
              data-testid="total-price"
            >
              <span
                class="font-medium"
              >
                $
              </span>
              10
              <span
                class="text-lg"
              >
                .
                00
              </span>
            </p>
          </div>
          <button
            class="mantine-Button-default mantine-Button-root text-gray-800 mantine-aywjzy"
            type="button"
          >
            <div
              class="mantine-3xbgk5 mantine-Button-inner"
            >
              <span
                class="mantine-qo1k2 mantine-Button-label"
              >
                Place Order 
                <svg
                  aria-hidden="true"
                  class="w-5 h-5 ml-1"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M14 5l7 7m0 0l-7 7m7-7H3"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
